File: aula04/int2bin-fast.f90
! Este programa tem exatamente o mesmo objetivo do programa "int2bin.f90"
! mas realiza uma operação de comparação bit a bit ao invés de um cálculo de módulo
! e uma operação de deslocamento de bit ao invés de uma divisão.
! Como as operações utilizadas aqui são computacionalmente mais baratas, este programa é mais eficiente.
! Além disso, este programa funciona com inteiros negativos.

PROGRAM int2bin

  ! Declaração das variáveis
  IMPLICIT NONE
  INTEGER(KIND=4) :: input, index

  WRITE(*,*) "Insira um númetro inteiro para ser convertido"
  READ(*,*) input

  ! Converte o valor em binário e imprime o resultado na saída pradrão
  DO index=1,32
    WRITE(*,*) AND(input, 1)
    input = RSHIFT(input, 1)
  END DO

END PROGRAM int2bin

File: aula04/int2bin-fast.cpp
using namespace std;

#include <iostream>
#include <stdlib.h>

// Este programa tem exatamente o mesmo objetivo do programa "conversor-decimal-binario"
// mas realiza uma operação de comparação bit a bit ao invés de um cálculo de módulo
// e uma operação de deslocamento de bit ao invés de uma divisão.
// Como as operações utilizadas aqui são computacionalmente mais baratas, este programa é mais eficiente.
// Além disso, este programa funciona com inteiros negativos.
int main(int argc, char const *argv[]) {

	int decimal = atoi(argv[1]);
	bool binario[32];

	for(int i = 0; i < 32; i++){
		binario[i] = (decimal&1);
		cout << binario[i];
		decimal >>= 1;
	}

	cout << endl;
	return 0;
}
